src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -779,6 +780,9 @@ address SharedRuntime::compute_compiled_exc_handler(CompiledMethod* cm, address
     return nullptr;
   }
 
+  if (handler_bci != -1) { // did we find a handler in this method?
+    sd->method()->set_ex_handler_entered(handler_bci); // profile
```

Comment:
Are you sure this handles inlined exception handlers properly for c2?  My recollection is that C2 generates single level tables and performs the inlined exception dispatch in the generated code.  If you notice the flag top_frame_only stops the walk through the callers and OptoRuntime::handle_exception_C always passes true for top_frame_only.  So it seems like sd will always be the ScopeDesc of the top frame.

Reply:
Hmm, you're right. The current code doesn't seem to work for exceptions thrown in an inlinee and caught in a caller. The other way around works.

I thought we only say `-1` when we had to unwind a frame, and then we'd do the exception handler lookup again. But, looking now, there doesn't seem to be a following lookup.

Reply:
If the expectation is that unhandled exception entry points will be deopts, then you may not actually have to do anything since the exception handling would end up going through the interpreter.  Or maybe Deoptimization::uncommon_trap_inner could be augmented to update the MDO at the right point?

Reply:
Thinking more about this: marking the handler as entered is mostly useful for C2 so that when we deoptimize and compile in C2 again, we don't emit an uncommon trap again. However, that would also require not seeing any exception in the interpreter or in tier 1-3. (Maybe not tracking C2 handler enters is actually desirable).

I think we can mark the handler as entered in the deoptimization code that handles the uncommon trap that replaces the ex. handler instead, for C2.

Reply:
> If the expectation is that unhandled exception entry points will be deopts, then you may not actually have to do anything since the exception handling would end up going through the interpreter.

The uncommon trap is located in the exception handler itself, so after the deopt we start interpreting in the exception handler, and don't go through the interpreter's exception dispatch code (`InterpreterRuntime::exception_handler_for_exception`).

I think handling this in `Deoptimization::uncommon_trap_inner` can work though.

Reply:
Yes I think a solution like that would work for JVMCI as well.  C2 and JVMCI exception dispatch are very similar.